lib/semantics/semantics.cc

```diff
@@ -24,13 +24,35 @@
 #include "scope.h"
 #include "symbol.h"
 #include "../common/default-kinds.h"
+#include "../parser/parse-tree-visitor.h"
 #include <ostream>
 
 namespace Fortran::semantics {
 
 static void DoDumpSymbols(std::ostream &, const Scope &, int indent = 0);
 static void PutIndent(std::ostream &, int indent);
 
+// A parse tree visitor that calls Enter/Leave functions from each checker
+// class C supplied as template parameters. Enter is called before the node's
+// children are visited, Leave is called after. No two checkers may have the
+// same Enter or Leave function. Each checker must be constructible from
+// SemanticsContext and have BaseChecker as a virtual base class.
+template<typename... C> struct SemanticsVisitor : public virtual C... {
+  using C::Enter...;
+  using C::Leave...;
+  using BaseChecker::Enter;
+  using BaseChecker::Leave;
+  SemanticsVisitor(SemanticsContext &context) : C{context}... {}
+  template<typename N> bool Pre(const N &node) {
+    Enter(node);
```

Comment:
Does the "No two checkers may have the same Enter or Leave function" limitation implies that two checkers that belongs to the same visitor cannot perform their checks on a same node?
Is it only a technical limitations or is it also a wish to avoid potential conflicts between the checkers?
If it is only a technical limitations, you could with a few changes call `(...,C::Enter(node));` instead of `Enter(node);` so that you can call multiple checkers on the same node.

Reply:
The limitation is intentional, with the hope that it will help avoid conflicts. We will have to reconsider when we encounter the first case where two checkers want to do something on the same node.

We can easily add multiple passes of checkers when that happens, or relax the constraint as you suggest.

Another thing that might lead to more than one pass is the fact that expression checking is (at least for now) included in the first pass. If there were to be a checker that needed an expression to be analyzed before the `Pre` call on the `Expr` node, it would need to be in a later pass.

Reply:
OK thanks for detailing, I am fine with this conservative approach.